packages/loopring_v3/circuit/Utils/Data.h

```diff
@@ -222,353 +340,371 @@ class Ring
+    create.newWalletHash = ethsnarks::FieldT(j["newWalletHash"].get<std::string>().c_str());
 }
 
-class OnchainWithdrawal
+
+class OwnerChange
 {
 public:
-    ethsnarks::FieldT amountRequested;
-    BalanceUpdate balanceUpdate;
-    AccountUpdate accountUpdate;
+    ethsnarks::FieldT accountID;
+    ethsnarks::FieldT feeTokenID;
+    ethsnarks::FieldT fee;
+    ethsnarks::FieldT newOwner;
 };
 
-static void from_json(const json& j, OnchainWithdrawal& withdrawal)
+static void from_json(const json& j, OwnerChange& change)
 {
-    withdrawal.amountRequested = ethsnarks::FieldT(j.at("amountRequested").get<std::string>().c_str());
-    withdrawal.balanceUpdate = j.at("balanceUpdate").get<BalanceUpdate>();
-    withdrawal.accountUpdate = j.at("accountUpdate").get<AccountUpdate>();
+    change.accountID = ethsnarks::FieldT(j.at("accountID"));
+    change.feeTokenID = ethsnarks::FieldT(j.at("feeTokenID"));
+    change.fee = ethsnarks::FieldT(j["fee"].get<std::string>().c_str());
+    change.newOwner = ethsnarks::FieldT(j["newOwner"].get<std::string>().c_str());
 }
 
-class OnchainWithdrawalBlock
+class Transfer
 {
 public:
 
-    ethsnarks::FieldT exchangeID;
-
-    ethsnarks::FieldT merkleRootBefore;
-    ethsnarks::FieldT merkleRootAfter;
-
-    ethsnarks::LimbT startHash;
-
-    ethsnarks::FieldT startIndex;
-    ethsnarks::FieldT count;
-
-    std::vector<Loopring::OnchainWithdrawal> withdrawals;
+    ethsnarks::FieldT fromAccountID;
+    ethsnarks::FieldT toAccountID;
+    ethsnarks::FieldT tokenID;
+    ethsnarks::FieldT amount;
+    ethsnarks::FieldT feeTokenID;
+    ethsnarks::FieldT fee;
+    ethsnarks::FieldT validUntil;
+    ethsnarks::FieldT to;
+    ethsnarks::FieldT dualAuthorX;
+    ethsnarks::FieldT dualAuthorY;
+    ethsnarks::FieldT data;
+    ethsnarks::FieldT payerToAccountID;
+    ethsnarks::FieldT payerTo;
+    ethsnarks::FieldT payeeToAccountID;
+    ethsnarks::FieldT type;
 };
 
-static void from_json(const json& j, OnchainWithdrawalBlock& block)
+static void from_json(const json& j, Transfer& transfer)
+{
+    transfer.fromAccountID = ethsnarks::FieldT(j.at("fromAccountID"));
+    transfer.toAccountID = ethsnarks::FieldT(j.at("toAccountID"));
+    transfer.tokenID = ethsnarks::FieldT(j.at("tokenID"));
+    transfer.amount = ethsnarks::FieldT(j["amount"].get<std::string>().c_str());
+    transfer.feeTokenID = ethsnarks::FieldT(j.at("feeTokenID"));
+    transfer.fee = ethsnarks::FieldT(j["fee"].get<std::string>().c_str());
+    transfer.validUntil = ethsnarks::FieldT(j.at("validUntil"));
+    transfer.to = ethsnarks::FieldT(j["to"].get<std::string>().c_str());
+    transfer.dualAuthorX = ethsnarks::FieldT(j["dualAuthorX"].get<std::string>().c_str());
+    transfer.dualAuthorY = ethsnarks::FieldT(j["dualAuthorY"].get<std::string>().c_str());
+    transfer.data = ethsnarks::FieldT(j["data"].get<std::string>().c_str());
+    transfer.payerToAccountID = ethsnarks::FieldT(j.at("payerToAccountID"));
```

Comment:
what are payer/payee here?

Reply:
The values as provided by the payer of the transfer and the values provided by the payee (which can differ if it's an open transfer). The payer values are signed by the payer, the payee values are signed with the DA keys when DA is used. In `TransferCircuit.h` there's some pseudo code that explains the logic for them:

```
if (payer_to != 0) {
    require(payer_to == to);
    require(payer_toAccountID == payee_toAccountID);
}
// Allow sending to an address instead of a specific accountID.
if (payee_toAccountID != 0) {
    require(payee_toAccountID == toAccountID);
}
require (to != 0);
```